File: LeetCode/remote/count_complete_tree_nodes.rb
# https://leetcode.com/problems/count-complete-tree-nodes/

require 'test/unit'
extend Test::Unit::Assertions

class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val, left, right)
        @val = val
        @left, @right = left, right
    end
end

def count_nodes(root)
    left = measure(root, :left)
    right = measure(root, :right)
    return 2 ** left - 1 if left == right
    max = left
    diff = 0
    current = root
    while current
        diff += 1
        current = current.left
        left -= 1
        right = measure(current, :right)
        return 2 ** max - 1 - diff if left == right
    end

    return -1
end

def measure(root, direction)
    current = root
    count = 0
    while current
        count += 1
        case direction
            when :left then current = current.left
            when :right then current = current.right
        end
    end

    return count
end

root = TreeNode.new(4,
                    TreeNode.new(2,
                                 TreeNode.new(1, nil, nil),
                                 TreeNode.new(3, nil, nil)),
                    TreeNode.new(5,
                                 TreeNode.new(6, nil, nil), nil))

assert_equal(6, count_nodes(root))

File: LeetCode/remote/count_complete_tree_nodes.cs
//  https://leetcode.com/problems/count-complete-tree-nodes/
//
//  Given a complete binary tree, count the number of nodes.
//
//  Definition of a complete binary tree from Wikipedia:
//  In a complete binary tree every level, except possibly the last,
//  is completely filled, and all nodes in the last level are as far left as possible.
//  It can have between 1 and 2h nodes inclusive at the last level h.

//  https://leetcode.com/submissions/detail/52956033/
//
//
//  Submission Details
//  18 / 18 test cases passed.
//      Status: Accepted
//      Runtime: 320 ms
//          
//          Submitted: 4 minutes ago
//

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int CountNodes(TreeNode root) {
        var left = LeftDepth(root);
        var right = RightDepth(root);
        if (left == right)
        {
            return (1 << left) - 1;
        }
        else
        {
            return 1 + CountNodes(root.left) + CountNodes(root.right);
        }
    }
    
    public int LeftDepth(TreeNode root)
    {
        var i = 0;
        while (root != null)
        {
            i++;
            root = root.left;
        }
        
        return i;
    }
    
    public int RightDepth(TreeNode root)
    {
        var i = 0;
        while (root != null)
        {
            i++;
            root = root.right;
        }
        
        return i;
    }

    //  https://leetcode.com/submissions/detail/65670291/
    //
    //  Submission Details
    //  18 / 18 test cases passed.
    //      Status: Accepted
    //      Runtime: 524 ms
    //          
    //          Submitted: 0 minutes ago
    public int CountNodes(TreeNode root) {
        if (root == null) {
            return 0;
        }
        
        var leftHeight = Height(root, true);
        var rightHeight = Height(root, false);
        return leftHeight == rightHeight ? 
                    (1 << leftHeight) - 1 :
                    1 + CountNodes(root.left) + CountNodes(root.right);
    }
    
    public int Height(TreeNode root, bool left) {
        return root == null ? 0 :
                    1 + Height(left ? root.left : root.right, left);
    }
}
